airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/streams.py

```diff
@@ -0,0 +1,448 @@
+        res["page"] = (next_page_token or {}).get("next_page") or 1
+        return res
+
+
+class FullRefreshBasicStream(IncrementalBasicUnsortedPageStream, ABC):
+    """"Basic stream for endpoints where there are not any created_at or updated_at fields"""
+
+    state_checkpoint_interval = None
+    cursor_field = []
+
+
+class IncrementalBasicSortedCursorStream(IncrementalBasicUnsortedStream, ABC):
+    """basic stream for loading sorting data with cursor hashed pagination"""
+
+    def request_params(
+        self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+        res = super().request_params(stream_state, next_page_token)
+        self._save_cursor_state(stream_state)
+        res.update({"sort_by": self.cursor_field, "sort_order": "desc", "limit": self.state_checkpoint_interval})
+        before_cursor = (next_page_token or {}).get("before_cursor")
+        if before_cursor:
+            res["cursor"] = before_cursor
+        return res
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        if self.is_finished:
+            return None
+        before_cursor = response.json()["before_cursor"]
+
+        if before_cursor:
+            return {"before_cursor": before_cursor}
+        return None
+
+
+class IncrementalBasicSortedPageStream(IncrementalBasicUnsortedPageStream, ABC):
+    """basic stream for loading sorting data with normal pagination"""
+
+    def request_params(
+        self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+
+        self._save_cursor_state(stream_state)
+        res = {"sort_by": self.cursor_field, "sort_order": "desc", "limit": self.state_checkpoint_interval}
+
+        if (next_page_token or {}).get("before_cursor"):
+            res["cursor"] = next_page_token["before_cursor"]
+        return res
+
+
+class CustomTicketAuditsStream(IncrementalBasicSortedCursorStream, ABC):
+    """Custom class for ticket_audits logic because a data response has not standard struct"""
+
+    # ticket audits doesn't have the 'updated_by' field
+    cursor_field = "created_at"
+
+    # Root of response is 'audits'. As rule as an endpoint name is equel a response list name
+    response_list_name = "audits"
+
+
+class CustomCommentsStream(IncrementalBasicSortedPageStream, ABC):
+    """Custom class for ticket_comments logic because ZenDesk doesn't provide API
+    for loading of all comment by one direct endpoints. Thus at first we loads
+    all updated tickets and after this tries to load all created/updated comment
+    per every ticket"""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Flag of loaded state. it is tickets' loaging if it is False and
+        #  it is comments' loaging if it is vice versa
+        self._loaded = False
```

Comment:
probably it is better to have a separate classes for tickets and ticket_comments

Reply:
then from ticket_comments stream we can call ticket stream to get a list of tickets

Reply:
for example please check mixpanel - cohort member stream